frame/contracts/src/migration/v12.rs

```diff
@@ -0,0 +1,324 @@
+			let price_per_byte = T::DepositPerByte::get();
+			let price_per_item = T::DepositPerItem::get();
+			let bytes_before = module
+				.encoded_size()
+				.saturating_add(code.len())
+				.saturating_add(old::OwnerInfo::<T>::max_encoded_len()) as u32;
+			let items_before = 3u32;
+			let deposit_expected_before = price_per_byte
+				.saturating_mul(bytes_before.into())
+				.saturating_add(price_per_item.saturating_mul(items_before.into()));
+			let ratio = FixedU128::checked_from_rational(old.deposit, deposit_expected_before)
+				.unwrap_or_default()
+				.min(FixedU128::from_u32(1));
+			let bytes_after = code.len().saturating_add(CodeInfo::<T>::max_encoded_len()) as u32;
+			let items_after = 2u32;
+			let deposit_expected_after = price_per_byte
+				.saturating_mul(bytes_after.into())
+				.saturating_add(price_per_item.saturating_mul(items_after.into()));
+			let deposit = ratio.saturating_mul_int(deposit_expected_after);
+
+			let info = CodeInfo::<T> {
+				determinism: module.determinism,
+				owner: old.owner,
+				deposit,
+				refcount: old.refcount,
+			};
+
+			let amount = old.deposit.saturating_sub(info.deposit);
+			if !amount.is_zero() {
+				T::Currency::unreserve(&info.owner, amount);
+				log::debug!(
+					target: LOG_TARGET,
+					"Deposit refunded: {:?} Balance, to: {:?}",
+					&amount,
+					HexDisplay::from(&info.owner.encode())
+				);
+			} else {
+				log::warn!(
+					target: LOG_TARGET,
+					"new deposit: {:?} is more than old deposit: {:?}",
+					&info.deposit,
+					&old.deposit
+				);
+			}
+			CodeInfoOf::<T>::insert(hash, info);
+
+			self.last_key = Some(iter.last_raw_key().to_vec().try_into().unwrap());
+
+			(IsFinished::No, T::WeightInfo::v12_migration_step(code.len() as u32))
+		} else {
+			log::debug!(target: LOG_TARGET, "No more OwnerInfo to migrate");
+			(IsFinished::Yes, T::WeightInfo::v12_migration_step(0))
+		}
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade_step() -> Result<Vec<u8>, TryRuntimeError> {
+		let len = 5;
+		log::debug!(target: LOG_TARGET, "Taking sample of {} OwnerInfo(s)", len);
+		let sample: Vec<_> = old::OwnerInfoOf::<T>::iter()
+			.take(len)
+			.map(|(k, v)| {
+				let module = old::CodeStorage::<T>::get(k)
+					.expect("No PrefabWasmModule found for code_hash: {:?}");
+				let info: CodeInfo<T> = CodeInfo {
+					determinism: module.determinism,
+					deposit: v.deposit,
+					refcount: v.refcount,
+					owner: v.owner,
+				};
+				(k, info)
+			})
+			.collect();
+
+		let storage: u32 =
+			old::CodeStorage::<T>::iter().map(|(_k, v)| v.encoded_size() as u32).sum();
+		let mut deposit: BalanceOf<T> = Default::default();
+		old::OwnerInfoOf::<T>::iter().for_each(|(_k, v)| deposit += v.deposit);
+
+		Ok((sample, deposit, storage).encode())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade_step(state: Vec<u8>) -> Result<(), TryRuntimeError> {
+		let state = <(Vec<(CodeHash<T>, CodeInfo<T>)>, BalanceOf<T>, u32) as Decode>::decode(
```

Comment:
for readability, maybe destructure here, instead of using state.0, state.1, ...

Reply:
agreed. done.